projects/extension/src/background/index.ts

```diff
@@ -49,19 +56,40 @@ const getTime = () => {
   }${date.getSeconds()} ${date.getMilliseconds()}`
 }
 
+const maintainArr = (arr: logStructure[], incLog: logStructure): void => {
+  if (arr.length >= 1000) arr.shift()
+  arr.push(incLog)
+}
+
+// duplicate "Error" is intentional for the extreme case of level === 0
+const logLevels = ["Error", "Error", "Warn", "Info", "Debug"]
+
 const logger = (level: number, target: string, message: string) => {
-  const logLevels = ["Error", "Warn", "Info", "Debug"]
+  const incLog = {
+    time: getTime(),
+    level: logLevels[level],
+    target,
+    message,
+  }
+  const { error, warn, all } = logKeeper
   if (level !== 4) {
     // log all non-debug logs to background console
-    console.log(`${logLevels[level - 1]}: ${message}`)
-    if (level === 1) {
-      logKeeper.error.push(`${getTime()} | ${target} | ${message}`)
-    } else if (level === 2) {
-      logKeeper.warn.push(`${getTime()} | ${target} | ${message}`)
+    switch (level) {
+      case 0:
+      case 1:
+        maintainArr(error, incLog)
```

Comment:
```suggestion
  if (error.length >= 1000) error.shift()
  error.push(incLog)
```

Please just copy-paste this, there's no need for a separate function